sys/arm64/include/asm.h

```diff
@@ -57,19 +57,39 @@
 #define	PIC_SYM(x,y)	x
 #endif
 
-/* Alias for link register x30 */
+#define	INT_WIDTH	8
+#define	INTN(n)		n
+#define	INT(n)		x ## n
+
+#if __has_feature(capabilities)
+#define	CAP_WIDTH	16
+#define	CAPN(n)		c ## n
+#define	CAP(n)		c ## n
+#else
+#define	CAP_WIDTH	INT_WIDTH
+#define	CAPN(n)		INTN(n)
+#define	CAP(n)		INT(n)
+#endif
+
 #ifdef __CHERI_PURE_CAPABILITY__
-#define	REG_WIDTH	16
-#define	REGN(n)		c ## n
-#define	REG(n)		c ## n
+#define	PTR_WIDTH	CAP_WIDTH
+#define	PTRN(n)		CAPN(n)
+#define	PTR(n)		CAP(n)
+/* Alias for link register c30 */
 #define	clr		c30
 #else
-#define	REG_WIDTH	8
-#define	REGN(n)		n
-#define	REG(n)		x ## n
+#define	PTR_WIDTH	INT_WIDTH
+#define	PTRN(n)		INTN(n)
+#define	PTR(n)		INT(n)
+/* Alias for link register x30 */
 #define	lr		x30
```

Comment:
We could also add a `#define xsp sp` to allow PTR(sp)` to work? Then PTRN() will only be needed for special registers.

Reply:
Well, `sp` isn't an X register, it _is_ a special register, it just sometimes gets encoded as if it were, so that would be wrong. I don't see what's wrong with using `PTRN(sp)`.

Reply:
Yes it's not needed. I just though it might be nice since you can use PTR(zr)/PTR(lr), but not PTR(sp). And arguably zr is also a special register ;)

Reply:
The assembler treats `x31` and `xzr` as synonymous, and there's no `(x)zr` (set of) system register(s) like for `sp`, so it's special in the colloquial sense but not in the architectural sense.

Reply:
Also `PTR(lr)` doesn't work